### PR TITLE
Generalize non-const ClassInit lowering beyond only InitializeFrom insts

### LINKED ISSUE #5186: Crash lowering init of object with pointer and non-pointer members
### Description of the bug:

Splitting off from https://github.com/carbon-language/carbon-lang/issues/5094#issuecomment-2752106258
The following crashes during lowering:
```
class Other {
}

base class Base {
  var u: Other*;
  var v: Other;
}

var v: Other;

fn F1() {
  var u : Base = {
   .u = &v,
   .v = {}
  };
}
```
Paraphrasing https://github.com/carbon-language/carbon-lang/issues/5094#issuecomment-2752100964 here:

pretty unsure what's going on/how to make progress, but taking notes here to try to formulate some ideas and ultimately questions for those more in the know.

So, the crash happens with a pointer (say, initialized with pointing to a global) and another member initialized with `{}` and it happens, essentially, here: https://github.com/carbon-language/carbon-lang/blob/a7d9ac576c6ced9cf411fa9c75c78f6108e062a5/toolchain/lower/handle_aggregates.cpp#L142-L148
Because an element of the `class_init.elements` is a constant, but not an `InitializeFrom`, instead it's a `Converted`.

Without the second member, the elements are all non-constant.

Without the vtable, we don't reach `EmitAggregateInitializer` at all... ah, because the class_init is a constant, so goes down a different path, per: https://github.com/carbon-language/carbon-lang/blob/a7d9ac576c6ced9cf411fa9c75c78f6108e062a5/toolchain/lower/function_context.cpp#L95-L98

So the pointer initialization shouldn't be a constant constant, but maybe a template constant. 

I'll try some differential debugging to see if I can make more sense of how it's meant to work.

### PR DESCRIPTION
**Pull request by dwblaikie**
Fixes #5186

With @zygoloid's kind assistance, this generalizes the existing
non-const lowering of ClassInit, that had previously only handled
InitializeFrom, to find other cases - such as a nested ClassInit used to
initialize a class member.

This refactors the `FindReturnSlotArgForInitializer` from
`check/convert.cpp` into `sem_ir/file.{h,cpp}` for use from lowering
(since lower doesn't depend on check, which I assume is an intentional
layering constraint - so figured it made sense to move it to sem_ir, and
found one or two similar-ish utility functions in `sem_ir/file.{h,cpp}`,
so figured that was a good spot)

### CODE REVIEW

File: toolchain/lower/handle_aggregates.cpp [hunk 1]
```diff
@@ -141,9 +142,12 @@ static auto EmitAggregateInitializer(FunctionContext& context,
       // non-constant initialization.
       for (auto [i, ref_id] : llvm::enumerate(refs)) {
         if (context.sem_ir().constant_values().Get(ref_id).is_constant()) {
-          auto init =
-              context.sem_ir().insts().GetAs<SemIR::InitializeFrom>(ref_id);
-          HandleInst(context, ref_id, init);
```
Comment (danakj) [inline toolchain/lower/handle_aggregates.cpp:146]: Does HandleInst(... SemIR::InitializeFrom) still have callers, now that we're somewhat doing its job here inline?

File: toolchain/lower/handle_aggregates.cpp [hunk 2]
```diff
@@ -141,9 +142,12 @@ static auto EmitAggregateInitializer(FunctionContext& context,
...
+          auto dest_id =
+              SemIR::FindReturnSlotArgForInitializer(context.sem_ir(), ref_id);
```
Comment (danakj) [inline toolchain/lower/handle_aggregates.cpp:147]: This is a question for @zygoloid but I feel like we generally want the SemIR to express stuff and lower to be less clever. Should Check be ensuring that we end up with an InitializeFrom instruction as the output of Convert (or something above Convert) here?

File: toolchain/lower/handle_aggregates.cpp (hunk 1 above)
Comment (dwblaikie) [reply toolchain/lower/handle_aggregates.cpp:146]: Yeah, I think if the `InitializeFrom` is non-constant, then `HandleInst(... SemIR::InitializeFrom)` gets called from the direct emission path here:
```
Carbon::Lower::HandleInst(Carbon::Lower::FunctionContext&, Carbon::SemIR::InstId, Carbon::SemIR::InitializeFrom) ()
    at toolchain/lower/handle.cpp:182
LowerInstHelper<Carbon::SemIR::InitializeFrom> () at toolchain/lower/function_context.cpp:86
LowerInst () at ./toolchain/sem_ir/inst_kind.def:80
LowerBlockContents () at toolchain/lower/function_context.cpp:56
```

File: toolchain/lower/testdata/class/field.carbon [hunk 3]
```diff
@@ -59,16 +81,59 @@ fn Run() -> i32 {
...
+// CHECK:STDOUT: define void @main() !dbg !4 {
+// CHECK:STDOUT: entry:
+// CHECK:STDOUT:   %o.var = alloca {}, align 8, !dbg !7
+// CHECK:STDOUT:   %u.var = alloca { ptr, {} }, align 8, !dbg !7
+// CHECK:STDOUT:   call void @llvm.lifetime.start.p0(i64 0, ptr %o.var), !dbg !7
+// CHECK:STDOUT:   call void @llvm.lifetime.start.p0(i64 8, ptr %u.var), !dbg !7
+// CHECK:STDOUT:   %.loc15_33.2.a = getelementptr inbounds nuw { ptr, {} }, ptr %u.var, i32 0, i32 0, !dbg !8
+// CHECK:STDOUT:   store ptr %o.var, ptr %.loc15_33.2.a, align 8, !dbg !8
+// CHECK:STDOUT:   %.loc15_33.4.b = getelementptr inbounds nuw { ptr, {} }, ptr %u.var, i32 0, i32 1, !dbg !8
+// CHECK:STDOUT:   call void @llvm.memcpy.p0.p0.i64(ptr align 1 %.loc15_33.4.b, ptr align 1 @Other.val.loc15_33.5, i64 0, i1 false), !dbg !8
```
Comment (zygoloid) [inline toolchain/lower/testdata/class/field.carbon:131]: It'd be good to also include a test with a nonempty class Other to ensure we copy that properly too. (We ought to eventually elide emitting the memcpy call here because the size is a constant zero.)

File: toolchain/sem_ir/file.h [hunk 4]
```diff
@@ -391,6 +391,12 @@ enum class ExprCategory : int8_t {
 // Returns the expression category for an instruction.
 auto GetExprCategory(const File& file, InstId inst_id) -> ExprCategory;
 
+// Given an initializing expression, find its return slot argument. Returns
+// `None` if there is no return slot, because the initialization is not
+// performed in place.
+auto FindReturnSlotArgForInitializer(const File& sem_ir, InstId init_id)
+    -> InstId;
+
```
Comment (zygoloid) [inline toolchain/sem_ir/file.h]: I think we should split these two functions and the `ExprCategory` enum out into their own file. Maybe `expr_info.h` / `.cpp`?

File: toolchain/lower/handle_aggregates.cpp (hunk 2 above)
Comment (zygoloid) [reply toolchain/lower/handle_aggregates.cpp:147]: Check is ensuring that we end up with an initializing expression here; handling an arbitrary initializing expression here seems sort of correct. `InitializeFrom` is only really needed in the case where there isn't another way to do in-place initialization (no other instruction to consume the `DestInstId` and ensure that it is stored to). I wonder on that basis if perhaps we're creating them more often than we need to. In any case, I think we'd benefit from more crisply specifying the intended semantics here.

I'm thinking of this function like `GetExprCategory`, `GetValueRepr`, etc. -- properties that are not stored directly in SemIR but somewhat trivially derived from it. `GetExprCategory` is used by the instruction formatter to print the `ref` / `init` marker on an instruction, and this function should be equivalent to what the formatter uses to print the `to %dest` part; things that are trivial enough to include in the formatted SemIR output seem like they probably ought to be acceptable and simple enough for lowering to use.

File: toolchain/sem_ir/file.h (hunk 4 above)
Comment (dwblaikie) [reply toolchain/sem_ir/file.h]: Ah, sure - 87e6b73495fd8322e2adf6576a644bb71ed8b20e

File: toolchain/lower/testdata/class/field.carbon (hunk 3 above)
Comment (dwblaikie) [reply toolchain/lower/testdata/class/field.carbon:131]: Given the emptiness isn't significant to the test - how about making this test be non-empty? Or do you think there's value in both empty and non-empty coverage?

This still reproduces the crash/demonstrates the fix:
```
class Other { var v: i32; }

class Use {
  var a: Other*;
  var b: Other;
}

fn Run() {
  var o: Other;
  var u: Use = {.a = &o, .b = {.v = 42}};
}
```

Comment (zygoloid) [reply toolchain/lower/testdata/class/field.carbon:131]: It might be good to test both the empty and non-empty case, because those code paths might reasonably diverge in the future, but I'd also be happy with testing only the non-empty case.

**Review by zygoloid — COMMENTED** [review]
LGTM other than wanting a test with a non-empty field; leaving to @danakj to approve once happy.

File: toolchain/lower/testdata/class/field.carbon (hunk 3 above)
Comment (dwblaikie) [reply toolchain/lower/testdata/class/field.carbon:131]: Sure thing! Added coverage for non-empty base in 41e190a6d2ce2e80e316d9d79198dd30999d6f38 so there's both non-empty and empty coverage.

**Review by danakj — APPROVED** [review]
LGTM
